bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java

```diff
@@ -525,43 +570,44 @@ public void startDialog(@Nullable KSService ks, @Nullable STTService stt, @Nulla
             @Nullable Voice voice, List<HumanLanguageInterpreter> hlis, @Nullable AudioSource source,
             @Nullable AudioSink sink, @Nullable Locale locale, @Nullable String keyword, @Nullable String listeningItem)
             throws IllegalStateException {
-        var context = getDefaultDialogContext();
+        var builder = getDialogContextBuilder();
         if (ks != null) {
-            context = context.withKS(ks);
+            builder.withKS(ks);
         }
         if (keyword != null) {
-            context = context.withKeyword(keyword);
+            builder.withKeyword(keyword);
         }
         if (stt != null) {
-            context = context.withSTT(stt);
+            builder.withSTT(stt);
         }
         if (tts != null) {
-            context = context.withTTS(tts);
+            builder.withTTS(tts);
         }
         if (voice != null) {
-            context = context.withVoice(voice);
+            builder.withVoice(voice);
         }
         if (!hlis.isEmpty()) {
-            context = context.withHLIs(hlis);
+            builder.withHLIs(hlis);
         }
         if (source != null) {
-            context = context.withSource(source);
+            builder.withSource(source);
         }
         if (sink != null) {
-            context = context.withSink(sink);
+            builder.withSink(sink);
         }
         if (locale != null) {
-            context = context.withLocale(locale);
+            builder.withLocale(locale);
         }
         if (listeningItem != null) {
-            context = context.withListeningItem(listeningItem);
+            builder.withListeningItem(listeningItem);
         }
-        startDialog(context);
+        startDialog(builder);
     }
 
     @Override
-    public void startDialog(DialogContext context) throws IllegalStateException {
+    public void startDialog(DialogContext.Builder contextBuilder) throws IllegalStateException {
         // use defaults, if null
+        DialogContext context = contextBuilder.build();
```

Comment:
Why do you pass the builder and not a context in `startDialog` and `stopDialog` when the very first thing that is done is creating a `DialogContext` with the builder? :thinking:

Reply:
Thinking on it, maybe for the method listenAndAnswer this makes no sense, as that one is invokes an immediate execution, I can change that one and let the builder just for the stopDialog and startDialog.

Reply:
> I think the way the voice manager handles the dialogs can be improved in the future, I was thinking on adding an issue after this one to gather some feedback. My idea is that the startdialog method can accept the contextBuilder and instead of failing if the services are not present it can register this dialog in a 'pending' state. 

I don't know what that means but it does not sound like it will make it into 3.4 for which the code freeze is in 2 days. So it is probably better to just pass an actual `DialogContext`. Also we are free to break and cleanup APIs with OH4. :wink:

Reply:
Sure, makes sense. Thank you for warning me about the code freeze I was not aware.

Reply:
The time line was announced here:

https://community.openhab.org/t/openhab-3-4-milestone-builds/138092/4

So we prefer to only merge bug fixes next week. 🙂

Reply:
Appreciated! I read that topic once in a while but I didn't know the timelines were announced there.